ros/src/twist_controller/twist_controller_mod.py

```diff
@@ -30,52 +31,61 @@ def __init__(self, wheel_base, steer_ratio, min_speed, max_lat_accel, max_steer_
+        # compute timestep
+        timestep = self.compute_timestep()
+        # use velocity pid controller to compute desired acceleration
+        acceleration_command = self.compute_acceleration_command(
+            target_linear_velocity - current_linear_velocity, timestep)
+        rospy.logwarn('Current linear velocity %.2f, target linear velocity %.2f, desired total acceleration %.2f',
+                      current_linear_velocity, target_linear_velocity, acceleration_command)
 
         # Return throttle, brake, steer
-        return (acceleration if acceleration > 0. else 0.,
-                self.compute_brake_torque(abs(acceleration)) if acceleration < 0. else 0.,
+        # TODO - build "coasting" to reduce speed rather than always using positive throttle or brake
+        return (self.compute_throttle(acceleration_command, target_linear_velocity, timestep) if acceleration_command > 0. else 0.,
+                self.compute_brake_torque(abs(acceleration_command)) if acceleration_command < 0. else 0.,
                 self.yaw_controller.get_steering(target_linear_velocity, target_angular_velocity, current_linear_velocity))
 
     def reset(self):
         self.last_timestep = None
         self.velocity_pid_controller.reset()
+        self.acceleration_pid_controller.reset()
 
-    def compute_acceleration(self, error):
+    def compute_timestep(self):
         last_timestep = self.current_timestep
-        self.current_timestep = int(round(time.time() * 1000))
-
-        if (last_timestep == self.current_timestep):
-            rospy.loginfo(
-                'No timestep advance since last acceleration computation; returning previous acceleration')
-            return self.previous_acceleration
+        self.current_timestep = time.time()
+        if last_timestep == None:
+            last_timestep = self.current_timestep - self.default_update_interval
+        return self.current_timestep - last_timestep
 
-        # TODO - modify PID instance to modify error value to prevent sudden control jumps when re-engaging DBW
-        # after a period of inactivity
-        if last_timestep is None:
-            # attempted to compute values for first time; do not accellerate
-            return 0.
+    def compute_acceleration_command(self, velocity_error, timestep):
+        return self.velocity_pid_controller.step(velocity_error, timestep)
 
-        self.previous_acceleration = self.velocity_pid_controller.step(
-            error, self.current_timestep - last_timestep)
-        return self.previous_acceleration
+    def compute_throttle(self, acceleration, target_linear_velocity, timestep):
+        # compute throttle based on desired acceleration plus a small amount of
+        # additional acceleration to combat wind resistance
+        return self.acceleration_pid_controller.step(acceleration + target_linear_velocity * self.acceleration_cruise_constant, timestep)
```

Comment:
If I understood correctly, changing the accel PID input from the error between target and actual accel to just the command accel (actually the proportional gained velocity error) plus another gain on the target speed is effectively rescaling the velocity feedback and adding a feed-forward factor.  Since this input is not an error term that we want to drive to zero, the accel PID controller is not actually doing a second feedback loop, so that should be why using any Integral gain component would not be stable.

Previous:
- velocity error * P feedback -> (accel target - actual accel) = accel error * PI feedback -> throttle
- accel target * mr -> brake

New:
- velocity error * P feedback -> (accel target + feedforward term (velocity * cruise gain)) * accel gain -> throttle
- accel target * mr -> brake

I think this does avoid the noise of trying to estimate actual acceleration from differentiating velocity and also allows having different FB gains for brake and throttle (brake is only affected by velocity P gain, throttle is affected by velocity P gain * accel P gain), and also adds feed-forward compensation for road load drag (velocity * cruise gain) to reduce the amount of feedback needed, but may not converge to the actual target speed if there is any variation in actual load without any integral feedback.  I think the feed-forward cruise gain can be tuned to get the car to track target speeds in the simulation since it's always repeatable, but may have more error in the actual car or during driving dynamics like tracking a moving speed target during accels/decels.  Also tuning the gains for throttle/brake response may be tricky since they are sequentially coupled.

The dual PID architecture I was thinking of would be something like:
- feedforward term (velocity * cruise gain) + velocity error * PI feedback -> throttle command
- velocity error * separate PI feedback (P gain can include mr multiplier) -> brake command
- throttle output = throttle command >= 0
- brake output = brake command if (throttle command == 0) and (actual velocity < target velocity - cruise margin (~5kph))

For now, we can merge this controller in to improve over the base, but let me know what you think about trying out the dual PID architecture I proposed above.

Reply:
Wow, I am so glad to have a tutor so knowledgeable in control systems! I'm very grateful for your detailed feedback and explanation. Thanks for putting up with my rudimentary knowledge and trial-and-error approach to improvements.

I would like to try your suggestion in a future piece of work. Thanks!